File: events/2015/13/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";
import { stringify } from "node:querystring";

function calculateHappiness(people: Record<string, Record<string, number>>, arrangement: string[]): number {
  let total = 0

  let len = arrangement.length;

  // Loop through the arrangement calculating the happiness for each each pair
  for (var i = 0; i < len; i++) {
    // Get the names of people, wrap around the array at the ends
    let person = arrangement[i];
    let left = arrangement[(i-1 + len) % len];
    let right = arrangement[(i+1) % len];

    total += people[person][left] + people[person][right];
  }

  return total
}

function generateAllPermutations(names: string[]): string[][] {
  // If the array only contains 1 item, return it
  if (names.length <= 1) { return [names]}

  // Start an output array
  let output: string[][] = [];

  // Pick each item of the array and recursivly run
  for (var i = 0; i < names.length; i++) {
    // Remove current item
    let rest = [...names.slice(0, i), ...names.slice(i + 1)]
    let permutations = generateAllPermutations(rest);

    for (const result of permutations) {
      output.push([names[i], ...result]);
    }
  }

  return output
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): number | undefined {
  let people: Record<string, Record<string, number>> = {}
  let names: string[] = [];

  for (var i = 0; i < input.length; i++) {
    if (input[i].length == 0) { continue }
    let parts = input[i].split(" ");
    let personA = parts[0];
    let personB = parts[10];
    personB = personB.substring(0, personB.length - 1)
    let change = parts[3];
    let posOrNeg = parts[2];

    if (people[personA] == undefined) {
      people[personA] = {}
    }

    if (posOrNeg == "gain") {
      people[personA][personB] = parseInt(change);
    } else if (posOrNeg == "lose") {
      people[personA][personB] = -parseInt(change);
    }

    if (!names.includes(personA)) {
      names.push(personA);
    }
  }

  // Pick the first person
  let first = names[0];
  let rest = [...names.slice(1)];
  let permutationsOfRest =  generateAllPermutations(rest);
  let allPermutations = permutationsOfRest.map((p) => [first, ...p]);

  let maxHappiness = -Infinity;
  for (const perm of allPermutations) {
    let hapiness = calculateHappiness(people, perm);
    if (hapiness > maxHappiness) {
      maxHappiness = hapiness
    }
  }

  return maxHappiness;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  let people: Record<string, Record<string, number>> = {}
  let names: string[] = [];

  for (var i = 0; i < input.length; i++) {
    if (input[i].length == 0) { continue }
    let parts = input[i].split(" ");
    let personA = parts[0];
    let personB = parts[10];
    personB = personB.substring(0, personB.length - 1)
    let change = parts[3];
    let posOrNeg = parts[2];

    if (people[personA] == undefined) {
      people[personA] = {}
    }

    if (posOrNeg == "gain") {
      people[personA][personB] = parseInt(change);
    } else if (posOrNeg == "lose") {
      people[personA][personB] = -parseInt(change);
    }

    if (!names.includes(personA)) {
      names.push(personA);
    }
  }

  // Add myself to the list
  people["Me"] = {}
  for (const person in people) {
    people[person]["Me"] = 0;
    people["Me"][person] = 0;
  }
  names.push("Me");

  // Pick the first person
  let first = names[0];
  let rest = [...names.slice(1)];
  let permutationsOfRest =  generateAllPermutations(rest);
  let allPermutations = permutationsOfRest.map((p) => [first, ...p]);

  let maxHappiness = -Infinity;
  for (const perm of allPermutations) {
    let hapiness = calculateHappiness(people, perm);
    if (hapiness > maxHappiness) {
      maxHappiness = hapiness
    }
  }

  return maxHappiness;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2015,
  Day: 13,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));

File: events/2015/13/solution.rs
use aoc_utils::read_input_as_array;
use aoc_utils::measure_performance;
use aoc_utils::AOCDayResults;
use aoc_utils::AOCTimestamp;
use chrono::{DateTime, Utc};
use std::collections::{HashMap, HashSet};

fn calculate_happiness(people: &HashMap<String, HashMap<String, i32>>, arrangement: &[String]) -> i32 {
  let mut total = 0;
  let n = arrangement.len();

  for i in 0..n {
    let left = &arrangement[(i + n - 1) % n];
    let right = &arrangement[(i + 1) % n];
    let person = &arrangement[i];

    if let Some(prefs) = people.get(person) {
      total += prefs.get(left).unwrap_or(&0);
      total += prefs.get(right).unwrap_or(&0);
    }
  }

  total
}

fn generate_all_permutations(names: &[String]) -> Vec<Vec<String>> {
  // If the array only contains 1 item, return it
  if names.len() <= 1 {
    return vec![names.to_vec()];
  }

  // Start an output array
  let mut output: Vec<Vec<String>> = Vec::new();

  for i in 0..names.len() {
    let mut rest = Vec::new();
    rest.extend_from_slice(&names[0..i]);
    rest.extend_from_slice(&names[i + 1..]);

    let permutations = generate_all_permutations(&rest);

    for result in permutations {
      let mut perm = vec![names[i].clone()];
      perm.extend(result);
      output.push(perm);
    }
  }

  output
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
fn part1(input: &Vec<String>) -> i32 {
  let mut people: HashMap<String, HashMap<String, i32>> = HashMap::new();
  let mut names_set: HashSet<String> = HashSet::new();

  for line in input {
    if line.is_empty() {
      continue;
    }

    let parts: Vec<&str> = line.split_whitespace().collect();
    let person_a = parts[0].to_string();
    let person_b = parts[10].trim_end_matches('.').to_string();
    let change: i32 = parts[3].parse().unwrap();
    let pos_or_neg = parts[2];

    people.entry(person_a.clone()).or_insert_with(HashMap::new);

    if pos_or_neg == "gain" {
      people.get_mut(&person_a).unwrap().insert(person_b, change);
    } else if pos_or_neg == "lose" {
      people.get_mut(&person_a).unwrap().insert(person_b, -change);
    }

    names_set.insert(person_a);
  }

  let names: Vec<String> = names_set.into_iter().collect();

  let first = names[0].clone();
  let rest: Vec<String> = names[1..].to_vec();
  let permutations_of_rest = generate_all_permutations(&rest);

  let all_permutations: Vec<Vec<String>> = permutations_of_rest
    .into_iter()
    .map(|mut p| {
      let mut result = vec![first.clone()];
      result.append(&mut p);
      result
    })
    .collect();

  let mut max_hapiness = i32::MIN;
  for perm in all_permutations {
    let happiness = calculate_happiness(&people, &perm);
    if happiness > max_hapiness {
      max_hapiness = happiness;
    }
  }

  max_hapiness
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
fn part2(input: &Vec<String>) -> i32 {
  let mut people: HashMap<String, HashMap<String, i32>> = HashMap::new();
  let mut names_set: HashSet<String> = HashSet::new();

  for line in input {
    if line.is_empty() {
      continue;
    }

    let parts: Vec<&str> = line.split_whitespace().collect();
    let person_a = parts[0].to_string();
    let person_b = parts[10].trim_end_matches('.').to_string();
    let change: i32 = parts[3].parse().unwrap();
    let pos_or_neg = parts[2];

    people.entry(person_a.clone()).or_insert_with(HashMap::new);

    if pos_or_neg == "gain" {
      people.get_mut(&person_a).unwrap().insert(person_b, change);
    } else if pos_or_neg == "lose" {
      people.get_mut(&person_a).unwrap().insert(person_b, -change);
    }

    names_set.insert(person_a);
  }

  let existing_names: Vec<String> = people.keys().cloned().collect();
  people.entry("Me".to_string()).or_insert_with(HashMap::new);

  for person in &existing_names {
    people.get_mut(person).unwrap().insert("Me".to_string(), 0);
    people.get_mut("Me").unwrap().insert(person.clone(), 0);
  }
  names_set.insert("Me".to_string());

  let names: Vec<String> = names_set.into_iter().collect();

  let first = names[0].clone();
  let rest: Vec<String> = names[1..].to_vec();
  let permutations_of_rest = generate_all_permutations(&rest);

  let all_permutations: Vec<Vec<String>> = permutations_of_rest
    .into_iter()
    .map(|mut p| {
      let mut result = vec![first.clone()];
      result.append(&mut p);
      result
    })
    .collect();

  let mut max_hapiness = i32::MIN;
  for perm in all_permutations {
    let happiness = calculate_happiness(&people, &perm);
    if happiness > max_hapiness {
      max_hapiness = happiness;
    }
  }

  max_hapiness
}

// ----------------------------------------------------------------------------------------------------
// | Main Function
// ----------------------------------------------------------------------------------------------------
fn main() {
  // Read the arguments
  let args = std::env::args().collect::<Vec<String>>();
  let default_file_name = String::from("input.txt");
  let file_name = args.get(1).unwrap_or(&default_file_name);

  // Start the timer
  let start_timestamp = std::time::SystemTime::now();
  let start_time = std::time::Instant::now();
  
  // Read the input
  let input = read_input_as_array(file_name).unwrap();

  // Run the Parts
  let p1_result = measure_performance(part1, &input);
  let p2_result = measure_performance(part2, &input);

  // Stop the timer
  let end_time = std::time::Instant::now();
  let end_timestamp = std::time::SystemTime::now();
  let duration = end_time.duration_since(start_time).as_nanos();

  let results = AOCDayResults {
    year: 2015,
    day: 13,
    part1: p1_result,
    part2: p2_result,
    duration: duration,
    timestamp: AOCTimestamp {
      start: DateTime::<Utc>::from(start_timestamp).to_rfc3339(),
      end: DateTime::<Utc>::from(end_timestamp).to_rfc3339(),
    },
  };

  let json_results = serde_json::to_string(&results).unwrap();
  println!("{}", json_results);
}